### PR TITLE
ContextScheduler for RxJava2 is added

### PR DESCRIPTION
**Pull request by SercanKaraoglu**
Signed-off-by: Sercan <sercan_karaoglu@yahoo.com>

### CODE REVIEW

**Comment by SercanKaraoglu** [comment]
@vietj here my PR is

**Comment by vietj** [comment]
@SercanKaraoglu it needs tests

**Comment by SercanKaraoglu** [comment]
Alright I just added them https://github.com/vert-x3/vertx-rx/pull/66/commits/792e38c9af7f0dba97a807634cb63290fb4afe1b @vietj

**Comment by vietj** [comment]
thanks, I'm currently rebasing this branch against the http-request-builder-branch and after that I'll merge your branch and push all

**Comment by SercanKaraoglu** [comment]
no problem, I know core team is pretty busy and let me know if you need hand for migrating to rxjava2 or things like that

**Comment by vietj** [comment]
@SercanKaraoglu any help is welcome, specially with respect to usability and respect of the RxJava API

I am working currently on improving the RxJava 1 API usability wise (also related to the new web client that is designed to work well with RxJava). So you can checkout these changes too (mainly introducing rxXXX methods instead of XXXObservable method that call the method every time on subscribe).

so any usability improvement that can be suggested or contributed is important. Right now I'm a bit lost by context switch between RxJava 1 improvements and RxJava 2 branch and I'm at least keeping this branch in sync with the master changes. I'll give thoughts about this - don't hesitate to join https://groups.google.com/forum/#!forum/vertx-dev where discussions have already happened and will likely continue.

**Comment by SercanKaraoglu** [comment]
I spent few days on using rxified api of http client and right it has some usability issues so i decided to leave rxified api and developed my own integration with rxjava2 also handled 30x redirections and no content 204 case. Currently I use that my own integration in production cases like doing thousands of requests and collecting results in one json object, using Single.merge function of rxjava2 currently code is not perfect but at least it is doing what I want it to do, I would be appreciated to help on that, http client is really important for us for now

**Comment by SercanKaraoglu** [comment]
It performs well for now but I think there is a problem with pipelining especially when connection is closed by server http client is having hard time to handle that and raise VertxException Connection was closed

**Comment by vietj** [comment]
@SercanKaraoglu yes but that is rather an HttpClient thing that should be improved.

btw I was speaking of https://github.com/vert-x3/vertx-web/tree/web-client

**Comment by vietj** [comment]
@SercanKaraoglu can you reformat your code to follow https://github.com/vert-x3/wiki/wiki/Vert.x-code-style-guidelines ?

**Comment by SercanKaraoglu** [comment]
@vietj I reformatted code using https://github.com/eclipse/vert.x/blob/master/vertx_code_style.xml please let me know if there is another issues

**Comment by vietj** [comment]
@SercanKaraoglu can you rebase on top of latest push to rx-java2 version ? in particular I've bumped the rxjava2 version and made the Single/Completable call the method on subscribe instead.

**Comment by SercanKaraoglu** [comment]
Nice, I did similar in my own work as well, I am collecting the response buffer into one buffer and return Single<Buffer> like this;
```java
public Single<Buffer> getWholeBody(String requestURI) {
        JsonArray urlsList = new JsonArray();
        JsonObject info = new JsonObject();
        info.put("url", requestURI);
        urlsList.add(info);
        return getBufferObservable(urlsList, info, requestURI)
                .doOnError(e -> LOGGER.error("{}", e));
    }
```

...I am rebasing this one.

**Comment by SercanKaraoglu** [comment]
done @vietj

File: rx-java-2/src/main/resources/vertx-reactivex/template/common.templ [hunk 1]
```diff
@@ -147,14 +147,6 @@
     }
     return expr;
   }
-
-  def capitalizeFirstLetter(s) {
```
Comment (vietj) [inline rx-java-2/src/main/resources/vertx-reactivex/template/common.templ]: why is this removed ?

File: rx-java-2/src/main/resources/vertx-reactivex/template/common.templ [hunk 2]
```diff
@@ -208,20 +200,12 @@
 	@code{futureParams=new java.util.ArrayList(method.params)}
 	@code{futureParam=futureParams.remove(futureParams.size()-1)}
 	@code{futureType=futureParam.type.args[0].args[0]}
-	@if{futureType.kind == CLASS_VOID}
-		@code{futureReturnType=io.vertx.codegen.type.TypeReflectionFactory.create(io.reactivex.Completable)}
```
Comment (vietj) [inline rx-java-2/src/main/resources/vertx-reactivex/template/common.templ]: why is this removed ?

File: .gitignore [hunk 3]
```diff
@@ -21,11 +21,9 @@ test-results
 test-tmp
 *.class
 src/gen
-rx-java/src/main/generated/**
```
Comment (vietj) [inline .gitignore]: why is this removed ?

File: .gitignore [hunk 4]
```diff
@@ -21,11 +21,9 @@ test-results
...
 rx-java/src/main/java/io/vertx/core/**
 rx-java/src/test/java/io/vertx/codegen/**
 rx-java/src/test/java/com/acme/**
-rx-java-2/src/main/generated/**
```
Comment (vietj) [inline .gitignore]: why is this removed ?

**Comment by SercanKaraoglu** [comment]
After rebase It is merged with this fork and I think that is when they are removed, let me check it out back again

**Comment by SercanKaraoglu** [comment]
@vietj can you please check it out It should be done by now

**Comment by vietj** [comment]
merged thanks

**Comment by vietj** [comment]
teh tests don't pass, did you run them ?

**Comment by SercanKaraoglu** [comment]
@vietj I could not because project wasnt compiling by the time I forked the branch because of some missing components, İf everything in place now I can look at where the problem is

**Comment by vietj** [comment]
yes please, I rolled back the merge

**Comment by SercanKaraoglu** [comment]
@vietj I still could not build project followings are missing
import io.vertx.reactivex.core.Future;
import io.vertx.reactivex.core.Context;
import io.vertx.reactivex.core.Vertx;
import io.vertx.reactivex.core.buffer.Buffer;
import io.vertx.reactivex.core.eventbus.EventBus;
import io.vertx.reactivex.core.eventbus.Message;
import io.vertx.reactivex.core.eventbus.MessageConsumer;
import io.vertx.reactivex.core.http.HttpClient;
import io.vertx.reactivex.core.http.HttpClientRequest;
import io.vertx.reactivex.core.http.HttpClientResponse;
import io.vertx.reactivex.core.http.HttpServer;
import io.vertx.reactivex.core.http.HttpServerRequest;
import io.vertx.reactivex.core.http.ServerWebSocket;
import io.vertx.reactivex.core.http.WebSocket;
import io.vertx.reactivex.core.net.NetServer;
import io.vertx.reactivex.core.net.NetSocket;
import io.vertx.codegen.testmodel.TestInterfaceImpl;
import io.vertx.core.json.JsonArray;
import io.vertx.reactivex.codegen.testmodel.TestInterface;

**Comment by SercanKaraoglu** [comment]
@vietj maybe test dependencies are not configured right in the pom.xml ?

**Comment by SercanKaraoglu** [comment]
@vietj ups sorry, I've just noticed bnd plugin

**Comment by SercanKaraoglu** [comment]
@vietj I fixed the failed tests, you can merge this time and let me know if there is something else.

**Comment by vietj** [comment]
can you squash your commits and make a new PR please ?

**Comment by SercanKaraoglu** [comment]
@vietj Could you look at the other PR where commits are squashed https://github.com/vert-x3/vertx-rx/pull/67

**Comment by vietj** [comment]
sure will do

**Comment by cescoffier** [comment]
Marked as invalid as it focus on RX2 (RX2 branch)
